File: Java/src/main/java/com/DSA/Java/Algorithms/DynamicProgramming/FibonacciWithDp.java
package com.DSA.Java.Algorithms.DynamicProgramming;

import java.util.ArrayList;
import java.util.HashMap;

public class FibonacciWithDp {
  private HashMap<Integer, Integer> hashMap;

  public FibonacciWithDp() {
    this.hashMap = new HashMap<>();
  }

  public int findNumOfFibWithMemo(int num) {
    if (num == 1) {
      return 0;
    }
    if (num == 2) {
      return 1;
    }
    if (!hashMap.containsKey(num)) {
      hashMap.put(num, (findNumOfFibWithMemo(num - 1) + findNumOfFibWithMemo(num - 2)));
    }
    return hashMap.get(num);
  }

  public int findNumOfFibWithTab(int num) {
    ArrayList<Integer> table = new ArrayList<>();
    table.add(0);
    table.add(1);
    for (int i = 2; i < num; i++) {
      int n1 = table.get(i - 1);
      int n2 = table.get(i - 2);
      table.add(n1 + n2);
    }
    return table.get(num - 1);
  }
}

File: C++/Algorithm/src/DynamicProgramming/FibonacciWithDP.cpp
#include "../../include/DynamicProgramming/FibonacciWithDP.h"

int FibonacciWithDP::fibonacci_problem_with_top_down_method(int num) {
  if (num == 1) return 0;
  if (num == 2) return 1;

  if (keys.find(num) == keys.end()) {
    keys[num] = fibonacci_problem_with_top_down_method(num - 1) + 
      fibonacci_problem_with_top_down_method(num - 2);
  }
  return keys[num];
}

int FibonacciWithDP::fibonacci_problem_with_bottom_up_method(int num) {
  std::vector<int> table {0, 1};
  for (int i = 2; i < num; i++) {
    int num_1 = table[i - 1];
    int num_2 = table[i - 2];
    table.push_back(num_1 + num_2);
  }
  return table[num - 1];
}
